File: app-desktop/src/test/kotlin/io/silencelen/andvari/desktop/ImportPushOutcomeTest.kt
package io.silencelen.andvari.desktop

import io.silencelen.andvari.core.client.ApiException
import io.silencelen.andvari.core.client.HouseholdCopy
import io.silencelen.andvari.core.client.KdfPolicyViolationException
import io.silencelen.andvari.core.client.UpgradeRequiredException
import io.silencelen.andvari.core.crypto.KdfParams
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * ux-error--1 (polish audit 2026-07-27): every import PUSH failure used to collapse into ONE
 * retryable sentence — "Import interrupted — press Retry to finish (no duplicates will be
 * created)." — so a 403, a weakened-KDF block, or a version pin replayed the same doomed push
 * forever under copy that promised it would finish. [importPushRetryable] / [importPushError] are
 * that split; android's twin pair is pinned by its PureGatesTest. The 426 escape (the catch-all
 * ALSO swallowed [UpgradeRequiredException], which drives the blocking upgrade screen everywhere
 * else) is control flow in importConfirm, asserted here as the contract this pair relies on.
 */
class ImportPushOutcomeTest {

    // ---- transient: the Retry affordance and its promise are honest ----

    @Test
    fun transportFailureKeepsTheRetryPromise() {
        val t = java.io.IOException("Connection reset by peer")
        assertTrue(importPushRetryable(t))
        assertEquals(IMPORT_INTERRUPTED, importPushError(t))
    }

    /** The server's own two retry-flavoured verdicts — its copy says "try again", so does ours. */
    @Test
    fun rateLimitAndServerHiccupStayRetryable() {
        assertTrue(importPushRetryable(ApiException(429, "rate_limited", "slow down")))
        assertTrue(importPushRetryable(ApiException(500, "internal", "boom")))
        assertTrue(importPushRetryable(ApiException(503, "unavailable", "restarting")))
        assertEquals(IMPORT_INTERRUPTED, importPushError(ApiException(500, "internal", "boom")))
    }

    /** An unclassified throw: the plan's itemIds double as push mutationIds, so the replay
     *  converges rather than duplicating — the one case where "no duplicates" is still true. */
    @Test
    fun unclassifiedThrowStaysRetryable() {
        assertTrue(importPushRetryable(IllegalStateException("mid-chunk")))
        assertEquals(IMPORT_INTERRUPTED, importPushError(IllegalStateException("mid-chunk")))
    }

    // ---- terminal: no Retry, and the copy names the refusal ----

    @Test
    fun permissionRefusalIsTerminalAndSaysSo() {
        val t = ApiException(403, "forbidden", "forbidden")
        assertFalse(importPushRetryable(t))
        assertEquals(HouseholdCopy.forError(t), importPushError(t))
        assertEquals("You don't have permission to do that.", importPushError(t))
    }

    @Test
    fun weakenedKdfPushIsTerminalAndKeepsTheSecurityWarning() {
        val t = KdfPolicyViolationException("kdf_below_floor", KdfParams())
        assertFalse(importPushRetryable(t))
        assertEquals(HouseholdCopy.WEAK_KDF_ACTION, importPushError(t))
    }

    @Test
    fun otherServerRefusalsAreTerminalWithCanonCopy() {
        for (t in listOf(
            ApiException(401, "unauthorized", "authentication failed"),
            ApiException(404, "not_found", "no such vault"),
            ApiException(409, "vault_state_changed", "rev mismatch"),
            ApiException(410, "gone", "purged"),
            ApiException(413, "too_large", "quota"),
        )) {
            assertFalse(importPushRetryable(t), "${t.status} replays as the same refusal")
            assertEquals(HouseholdCopy.forError(t), importPushError(t))
        }
    }

    /** The audited leak shape: a refusal's copy must never carry the server's raw message. */
    @Test
    fun terminalCopyNeverLeaksWireText() {
        val raw = "SQLITE_CONSTRAINT: UNIQUE constraint failed: mutations.id"
        val mapped = importPushError(ApiException(400, "bad_request", raw))
        assertFalse(mapped.contains("SQLITE"), "raw wire text leaked: $mapped")
    }

    /**
     * The 426 contract: a version pin must NOT be classified here at all — importConfirm catches
     * [UpgradeRequiredException] ahead of this pair and raises the blocking upgrade screen (op() /
     * runSync parity). This asserts the classifier would otherwise have called it terminal, so a
     * future refactor that drops that catch cannot quietly land back on a Retry prompt.
     */
    @Test
    fun upgradeRequiredIsNeverRetryable() {
        val t = UpgradeRequiredException("upgrade_required", "min version 0.22.0")
        assertFalse(importPushRetryable(t))
        assertEquals(HouseholdCopy.UPGRADE_REQUIRED, importPushError(t))
    }
}

File: web/src/ui/import-push-outcome.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { ApiError } from "../api/client";
import { KdfPolicyError, WEAK_KDF_MESSAGE } from "../crypto/keys";
import { NetworkError } from "./errors";
import { importPushOutcome } from "./Vault";

/**
 * ux-error--1 (polish audit 2026-07-27): every import PUSH failure collapsed into ONE retryable
 * sentence ("press Retry to import the rest"), so a permission refusal or a weakened-KDF block
 * replayed the same doomed push forever under copy that promised it would finish. The panel now
 * asks [importPushOutcome] whether the failure is terminal, and only a transient one keeps the
 * Retry button (the natives' twin split is pinned by desktop ImportPushOutcomeTest / android
 * PureGatesTest). Pure function + real sources, no rendering — the offlineCopyModel idiom.
 */

const here = (p: string) => fileURLToPath(new URL(p, import.meta.url));
const householdKt = readFileSync(
  here("../../../core/src/commonMain/kotlin/io/silencelen/andvari/core/client/HouseholdCopy.kt"),
  "utf8",
);
/** The curated sentence on HouseholdCopy's `e.status == <status>` row (vault-copy.test.ts's helper). */
function canonStatusRow(status: number): string {
  const m = householdKt.match(new RegExp(`e\\.status == ${status} -> "([^"]+)"`));
  expect(m, `HouseholdCopy's ${status} row moved — update the pin`).not.toBeNull();
  return m![1]!;
}

describe("import push outcome — terminal vs transient", () => {
  it("a transport blip keeps the Retry affordance and its no-duplicates promise", () => {
    const o = importPushOutcome(new NetworkError());
    expect(o.fatal).toBe(false);
    expect(o.message).toContain("press Retry");
  });

  it("a rate limit and a 5xx stay retryable — the server's own copy says to try again", () => {
    for (const e of [new ApiError(429, "rate_limited", "slow down"), new ApiError(503, "unavailable", "restarting")]) {
      expect(importPushOutcome(e).fatal, `${e.status} should stay retryable`).toBe(false);
    }
  });

  it("an unclassified throw stays retryable (the idempotent replay converges)", () => {
    expect(importPushOutcome(new Error("boom")).fatal).toBe(false);
  });

  it("a 403 is TERMINAL and shows the canon 403 sentence, not a retry promise", () => {
    const o = importPushOutcome(new ApiError(403, "forbidden", "forbidden"));
    expect(o.fatal).toBe(true);
    expect(o.message).toBe(canonStatusRow(403)); // byte-twin of the natives' HouseholdCopy row
    expect(o.message).not.toContain("Retry");
  });

  it("a 413 is TERMINAL and shows the canon 413 sentence", () => {
    const o = importPushOutcome(new ApiError(413, "too_large", "quota"));
    expect(o.fatal).toBe(true);
    expect(o.message).toBe(canonStatusRow(413));
  });

  it("a weakened-KDF push is TERMINAL and keeps the security warning (H1, spec 05 T1)", () => {
    const o = importPushOutcome(new KdfPolicyError("kdf_below_floor", { v: 19, alg: "argon2id", ops: 1, memBytes: 1024 }));
    expect(o.fatal).toBe(true);
    expect(o.message).toBe(WEAK_KDF_MESSAGE);
    expect(o.message).not.toContain("Retry");
  });

  it("never surfaces the server's raw message", () => {
    const raw = "SQLITE_CONSTRAINT: UNIQUE constraint failed: mutations.id";
    for (const status of [400, 403, 413, 500]) {
      expect(importPushOutcome(new ApiError(status, "err", raw)).message).not.toContain("SQLITE");
    }
  });
});

describe("the Retry button is gated on the outcome, not merely on there being an error", () => {
  const vaultTsx = readFileSync(here("./Vault.tsx"), "utf8");

  it("Vault.tsx renders the exit — never Retry — for a fatal push failure", () => {
    // The audited defect was structural: `importErr ? <Retry> : <Import N>` had no terminal arm,
    // so ANY push failure lit the Retry button. The fatal arm must come FIRST.
    const fatalArm = vaultTsx.indexOf("{importErr && importFatal ? (");
    const retryArm = vaultTsx.indexOf(") : importErr ? (");
    expect(fatalArm, "the terminal arm is gone — a refusal would light Retry again").toBeGreaterThan(-1);
    expect(retryArm).toBeGreaterThan(fatalArm);
    // …and the terminal arm's button is the exit (onDone, so landed rows show), not a retry.
    expect(vaultTsx.slice(fatalArm, retryArm)).toContain("onClick={onDone}");
    expect(vaultTsx.slice(fatalArm, retryArm)).not.toContain("runImport");
  });
});
